surya/common/surya/encoder/__init__.py

```diff
@@ -133,6 +139,133 @@ def apply_rotary_pos_emb_flashatt(
     return q_embed, k_embed
 
 
+class Qwen2_5_VLVisionXLASdpaAttention(nn.Module):
+    def __init__(self, dim: int, num_heads: int = 16) -> None:
+        super().__init__()
+        self.num_heads = num_heads
+        self.qkv = nn.Linear(dim, dim * 3, bias=True)
+        self.proj = nn.Linear(dim, dim)
+        self.head_dim = dim // num_heads
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        cu_seqlens: torch.Tensor,
+        rotary_pos_emb: Optional[torch.Tensor] = None,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+    ) -> torch.Tensor:
+        bsz, seq_length = hidden_states.shape[0], hidden_states.shape[1]
+        q, k, v = (
+            self.qkv(hidden_states)
+            .reshape(bsz, seq_length, 3, self.num_heads, -1)
+            .permute(0, 2, 1, 3, 4)
+            .unbind(1)
+        )
+        if position_embeddings is None:
+            logger.warning_once(
+                "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
+                "through `rotary_pos_emb` (2D tensor of RoPE theta values), to using externally computed "
+                "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.54 `rotary_pos_emb` will be "
+                "removed and `position_embeddings` will be mandatory."
+            )
+            emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
+            cos = emb.cos()
+            sin = emb.sin()
+        else:
+            cos, sin = position_embeddings
+        q, k = apply_rotary_pos_emb_vision(q, k, cos, sin)
+
+        attention_mask = torch.zeros([bsz, 1, seq_length, seq_length], dtype=torch.bool)
```

Comment:
Minor - I'm not sure if this will be a speed boost, but we don't need to recompute these seq lens on every layer. Might make sense to compute once in the `model.encode_images` function and pass in the attention mask